File: backend/app/services/ontology_action_service.py
import json
from sqlalchemy.orm import Session
from app.models.ontology import (
    OntologyActionType, OntologyActionParameter, OntologyActionRule,
    OntologyActionExecution, OntologyObjectType, OntologyLinkType,
    OntologyObject, OntologyLink, OntologyFunction,
)
from app.services import ontology_object_service as obj_service


# ===== ActionType CRUD =====

def list_action_types(db: Session):
    return db.query(OntologyActionType).all()


def get_action_type(db: Session, id: int) -> OntologyActionType:
    at = db.get(OntologyActionType, id)
    if not at:
        raise ValueError(f"ActionType not found: {id}")
    return at


def create_action_type(db: Session, data: dict) -> OntologyActionType:
    at = OntologyActionType(**data)
    db.add(at)
    db.commit()
    db.refresh(at)
    return at


def update_action_type(db: Session, id: int, data: dict) -> OntologyActionType:
    at = get_action_type(db, id)
    for key, value in data.items():
        if value is not None or key in ("target_object_type_id", "trigger_config_json",
                                         "exception_config_json", "validation_rules_json"):
            setattr(at, key, value)
    db.commit()
    db.refresh(at)
    return at


def delete_action_type(db: Session, id: int):
    get_action_type(db, id)
    db.query(OntologyActionParameter).filter(OntologyActionParameter.action_type_id == id).delete()
    db.query(OntologyActionRule).filter(OntologyActionRule.action_type_id == id).delete()
    db.query(OntologyActionType).filter(OntologyActionType.id == id).delete()
    db.commit()


# ===== Parameter CRUD =====

def list_parameters(db: Session, action_type_id: int):
    return db.query(OntologyActionParameter).filter(
        OntologyActionParameter.action_type_id == action_type_id
    ).order_by(OntologyActionParameter.sort_order).all()


def add_parameter(db: Session, action_type_id: int, data: dict) -> OntologyActionParameter:
    get_action_type(db, action_type_id)
    param = OntologyActionParameter(action_type_id=action_type_id, **data)
    db.add(param)
    db.commit()
    db.refresh(param)
    return param


def update_parameter(db: Session, id: int, data: dict) -> OntologyActionParameter:
    p = db.get(OntologyActionParameter, id)
    if not p:
        raise ValueError(f"Parameter not found: {id}")
    for key, value in data.items():
        if value is not None or key in ("required", "sort_order"):
            setattr(p, key, value)
    db.commit()
    db.refresh(p)
    return p


def delete_parameter(db: Session, id: int):
    p = db.get(OntologyActionParameter, id)
    if not p:
        raise ValueError(f"Parameter not found: {id}")
    db.delete(p)
    db.commit()


# ===== Rule CRUD =====

def list_rules(db: Session, action_type_id: int):
    return db.query(OntologyActionRule).filter(
        OntologyActionRule.action_type_id == action_type_id
    ).order_by(OntologyActionRule.sort_order).all()


def add_rule(db: Session, action_type_id: int, data: dict) -> OntologyActionRule:
    get_action_type(db, action_type_id)
    rule = OntologyActionRule(action_type_id=action_type_id, **data)
    db.add(rule)
    db.commit()
    db.refresh(rule)
    return rule


def update_rule(db: Session, id: int, data: dict) -> OntologyActionRule:
    r = db.get(OntologyActionRule, id)
    if not r:
        raise ValueError(f"Rule not found: {id}")
    for key, value in data.items():
        if value is not None or key == "sort_order":
            setattr(r, key, value)
    db.commit()
    db.refresh(r)
    return r


def delete_rule(db: Session, id: int):
    r = db.get(OntologyActionRule, id)
    if not r:
        raise ValueError(f"Rule not found: {id}")
    db.delete(r)
    db.commit()


# ===== Execution History =====

def list_executions(db: Session, action_type_id: int):
    return db.query(OntologyActionExecution).filter(
        OntologyActionExecution.action_type_id == action_type_id
    ).order_by(OntologyActionExecution.executed_at.desc()).all()


# ===== Execute Action =====

def execute_action(db: Session, action_type_id: int, input_parameters: dict) -> OntologyActionExecution:
    execution = OntologyActionExecution(
        action_type_id=action_type_id,
        parameters_json=_to_json(input_parameters),
    )

    try:
        action_type = get_action_type(db, action_type_id)
        if action_type.status != "ACTIVE":
            raise ValueError(f"ActionType is not ACTIVE: {action_type.status}")

        # Validate required parameters
        params = db.query(OntologyActionParameter).filter(
            OntologyActionParameter.action_type_id == action_type_id
        ).order_by(OntologyActionParameter.sort_order).all()
        for p in params:
            if p.required and p.name not in input_parameters:
                raise ValueError(f"Missing required parameter: {p.name}")

        # Execute rules in order
        rules = db.query(OntologyActionRule).filter(
            OntologyActionRule.action_type_id == action_type_id
        ).order_by(OntologyActionRule.sort_order).all()
        rule_results = []

        for rule in rules:
            if rule.condition_json and rule.condition_json.strip():
                if not _evaluate_condition(rule.condition_json, input_parameters):
                    continue
            rule_result = _execute_rule(db, rule, input_parameters)
            rule_results.append(rule_result)

        execution.status = "SUCCESS"
        execution.result_json = _to_json({"ruleResults": rule_results})

    except Exception as e:
        execution.status = "FAILED"
        execution.error_message = str(e)
        execution.result_json = _to_json({"error": str(e)})

    db.add(execution)
    db.commit()
    db.refresh(execution)
    return execution


def _execute_rule(db: Session, rule: OntologyActionRule, params: dict) -> dict:
    result = {"ruleId": rule.id, "ruleType": rule.rule_type}

    try:
        resolved_props = _resolve_property_mappings(rule.property_mappings_json, params)

        if rule.rule_type == "CREATE_OBJECT":
            ot = db.query(OntologyObjectType).filter(
                OntologyObjectType.name == rule.target_object_type_name
            ).first()
            if not ot:
                raise ValueError(f"ObjectType: {rule.target_object_type_name}")
            obj = OntologyObject(object_type_id=ot.id, properties_json=_to_json(resolved_props))
            created = obj_service.create_object(db, obj)
            result["objectId"] = created.id
            result["status"] = "created"

        elif rule.rule_type == "MODIFY_OBJECT":
            obj_id = params.get("objectId")
            if obj_id is None:
                raise ValueError("MODIFY_OBJECT requires 'objectId' parameter")
            obj_service.update_object_by_id(db, int(obj_id), {"properties_json": _to_json(resolved_props)})
            result["objectId"] = int(obj_id)
            result["status"] = "modified"

        elif rule.rule_type == "DELETE_OBJECT":
            obj_id = params.get("objectId")
            if obj_id is None:
                raise ValueError("DELETE_OBJECT requires 'objectId' parameter")
            obj_service.delete_object_by_id(db, int(obj_id))
            result["objectId"] = int(obj_id)
            result["status"] = "deleted"

        elif rule.rule_type == "CREATE_LINK":
            lt = db.query(OntologyLinkType).filter(
                OntologyLinkType.name == rule.target_link_type_name
            ).first()
            if not lt:
                raise ValueError(f"LinkType: {rule.target_link_type_name}")
            link = OntologyLink(
                link_type_id=lt.id,
                source_object_id=int(params["sourceObjectId"]),
                target_object_id=int(params["targetObjectId"]),
            )
            created = obj_service.create_link(db, link)
            result["linkId"] = created.id
            result["status"] = "created"

        elif rule.rule_type == "DELETE_LINK":
            link_id = int(params["linkId"])
            obj_service.delete_link(db, link_id)
            result["linkId"] = link_id
            result["status"] = "deleted"

        else:
            raise ValueError(f"Unknown ruleType: {rule.rule_type}")

    except Exception as e:
        result["status"] = "failed"
        result["error"] = str(e)

    return result


def _evaluate_condition(condition_json: str, params: dict) -> bool:
    try:
        condition = json.loads(condition_json)
        field = condition.get("field")
        operator = condition.get("operator")
        value = condition.get("value")
        actual = params.get(field)
        if actual is None:
            return False
        if operator == "eq":
            return str(actual) == str(value)
        elif operator == "neq":
            return str(actual) != str(value)
        elif operator == "exists":
            return True
        return True
    except Exception:
        return True


def _resolve_property_mappings(mappings_json: str, params: dict) -> dict:
    if not mappings_json or not mappings_json.strip():
        return {}
    try:
        mappings = json.loads(mappings_json)
        resolved = {}
        for key, val in mappings.items():
            val_str = str(val)
            if val_str.startswith("$parameters."):
                param_name = val_str[len("$parameters."):]
                resolved[key] = params.get(param_name, "")
            else:
                resolved[key] = val
        return resolved
    except Exception:
        return {}


def _to_json(obj) -> str:
    try:
        return json.dumps(obj, ensure_ascii=False)
    except Exception:
        return "{}"

File: java/device-manager/src/main/java/com/cccode/devicemanager/ontology/service/OntologyActionService.java
package com.cccode.devicemanager.ontology.service;

import com.cccode.devicemanager.ontology.model.*;
import com.cccode.devicemanager.ontology.repository.*;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Service
public class OntologyActionService {

    private final OntologyActionTypeRepository      actionTypeRepo;
    private final OntologyActionParameterRepository paramRepo;
    private final OntologyActionRuleRepository      ruleRepo;
    private final OntologyActionExecutionRepository executionRepo;
    private final OntologyObjectTypeRepository      objectTypeRepo;
    private final OntologyLinkTypeRepository        linkTypeRepo;
    private final OntologyObjectService             objectService;
    private final OntologyFunctionRepository        functionRepo;

    private final ObjectMapper mapper = new ObjectMapper();

    public OntologyActionService(OntologyActionTypeRepository actionTypeRepo,
                                  OntologyActionParameterRepository paramRepo,
                                  OntologyActionRuleRepository ruleRepo,
                                  OntologyActionExecutionRepository executionRepo,
                                  OntologyObjectTypeRepository objectTypeRepo,
                                  OntologyLinkTypeRepository linkTypeRepo,
                                  OntologyObjectService objectService,
                                  OntologyFunctionRepository functionRepo) {
        this.actionTypeRepo = actionTypeRepo;
        this.paramRepo      = paramRepo;
        this.ruleRepo       = ruleRepo;
        this.executionRepo  = executionRepo;
        this.objectTypeRepo = objectTypeRepo;
        this.linkTypeRepo   = linkTypeRepo;
        this.objectService  = objectService;
        this.functionRepo   = functionRepo;
    }

    // ===== ActionType CRUD =====

    public List<OntologyActionType> listActionTypes() {
        return actionTypeRepo.findAll();
    }

    public OntologyActionType getActionType(Long id) {
        return actionTypeRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("ActionType not found: " + id));
    }

    public OntologyActionType createActionType(OntologyActionType at) {
        return actionTypeRepo.save(at);
    }

    public OntologyActionType updateActionType(Long id, OntologyActionType patch) {
        OntologyActionType at = getActionType(id);
        if (patch.getDisplayName() != null)     at.setDisplayName(patch.getDisplayName());
        if (patch.getDescription() != null)     at.setDescription(patch.getDescription());
        if (patch.getStatus() != null)          at.setStatus(patch.getStatus());
        // 新增字段
        at.setTargetObjectTypeId(patch.getTargetObjectTypeId());
        if (patch.getTriggerType() != null)     at.setTriggerType(patch.getTriggerType());
        at.setTriggerConfigJson(patch.getTriggerConfigJson());
        if (patch.getExceptionPolicy() != null) at.setExceptionPolicy(patch.getExceptionPolicy());
        at.setExceptionConfigJson(patch.getExceptionConfigJson());
        at.setValidationRulesJson(patch.getValidationRulesJson());
        return actionTypeRepo.save(at);
    }

    @Transactional
    public void deleteActionType(Long id) {
        getActionType(id);
        paramRepo.deleteByActionTypeId(id);
        ruleRepo.deleteByActionTypeId(id);
        actionTypeRepo.deleteById(id);
    }

    // ===== Parameter CRUD =====

    public List<OntologyActionParameter> listParameters(Long actionTypeId) {
        return paramRepo.findByActionTypeIdOrderBySortOrder(actionTypeId);
    }

    public OntologyActionParameter addParameter(Long actionTypeId, OntologyActionParameter param) {
        getActionType(actionTypeId);
        param.setActionTypeId(actionTypeId);
        return paramRepo.save(param);
    }

    public OntologyActionParameter updateParameter(Long id, OntologyActionParameter patch) {
        OntologyActionParameter p = paramRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Parameter not found: " + id));
        if (patch.getDisplayName() != null)    p.setDisplayName(patch.getDisplayName());
        if (patch.getDataType() != null)       p.setDataType(patch.getDataType());
        if (patch.getDefaultValue() != null)   p.setDefaultValue(patch.getDefaultValue());
        if (patch.getConstraintsJson() != null) p.setConstraintsJson(patch.getConstraintsJson());
        p.setRequired(patch.isRequired());
        p.setSortOrder(patch.getSortOrder());
        return paramRepo.save(p);
    }

    public void deleteParameter(Long id) {
        paramRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Parameter not found: " + id));
        paramRepo.deleteById(id);
    }

    // ===== Rule CRUD =====

    public List<OntologyActionRule> listRules(Long actionTypeId) {
        return ruleRepo.findByActionTypeIdOrderBySortOrder(actionTypeId);
    }

    public OntologyActionRule addRule(Long actionTypeId, OntologyActionRule rule) {
        getActionType(actionTypeId);
        rule.setActionTypeId(actionTypeId);
        return ruleRepo.save(rule);
    }

    public OntologyActionRule updateRule(Long id, OntologyActionRule patch) {
        OntologyActionRule r = ruleRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Rule not found: " + id));
        if (patch.getRuleType() != null)              r.setRuleType(patch.getRuleType());
        if (patch.getTargetObjectTypeName() != null)  r.setTargetObjectTypeName(patch.getTargetObjectTypeName());
        if (patch.getTargetLinkTypeName() != null)    r.setTargetLinkTypeName(patch.getTargetLinkTypeName());
        if (patch.getPropertyMappingsJson() != null)  r.setPropertyMappingsJson(patch.getPropertyMappingsJson());
        if (patch.getConditionJson() != null)         r.setConditionJson(patch.getConditionJson());
        r.setSortOrder(patch.getSortOrder());
        return ruleRepo.save(r);
    }

    public void deleteRule(Long id) {
        ruleRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Rule not found: " + id));
        ruleRepo.deleteById(id);
    }

    // ===== Execution History =====

    public List<OntologyActionExecution> listExecutions(Long actionTypeId) {
        return executionRepo.findByActionTypeIdOrderByExecutedAtDesc(actionTypeId);
    }

    // ===== Execute Action =====

    @Transactional
    public OntologyActionExecution executeAction(Long actionTypeId, Map<String, Object> inputParameters) {
        OntologyActionExecution execution = new OntologyActionExecution();
        execution.setActionTypeId(actionTypeId);
        execution.setParametersJson(toJson(inputParameters));

        try {
            OntologyActionType actionType = getActionType(actionTypeId);
            if (!"ACTIVE".equals(actionType.getStatus())) {
                throw new IllegalStateException("ActionType is not ACTIVE: " + actionType.getStatus());
            }

            // Validate required parameters
            List<OntologyActionParameter> params = paramRepo.findByActionTypeIdOrderBySortOrder(actionTypeId);
            for (OntologyActionParameter p : params) {
                if (p.isRequired() && !inputParameters.containsKey(p.getName())) {
                    throw new IllegalArgumentException("Missing required parameter: " + p.getName());
                }
            }

            // Execute rules in order
            List<OntologyActionRule> rules = ruleRepo.findByActionTypeIdOrderBySortOrder(actionTypeId);
            List<Map<String, Object>> ruleResults = new ArrayList<>();

            for (OntologyActionRule rule : rules) {
                // Evaluate condition
                if (rule.getConditionJson() != null && !rule.getConditionJson().isBlank()) {
                    if (!evaluateCondition(rule.getConditionJson(), inputParameters)) {
                        continue;
                    }
                }

                Map<String, Object> ruleResult = executeRule(rule, inputParameters);
                ruleResults.add(ruleResult);
            }

            // Trigger associated functions (isolated from main transaction)
            triggerFunctions(actionTypeId, ruleResults, inputParameters, execution);

            execution.setStatus("SUCCESS");
            execution.setResultJson(toJson(Map.of("ruleResults", ruleResults)));

        } catch (Exception e) {
            execution.setStatus("FAILED");
            execution.setErrorMessage(e.getMessage());
            execution.setResultJson(toJson(Map.of("error", e.getMessage())));
        }

        return executionRepo.save(execution);
    }

    private Map<String, Object> executeRule(OntologyActionRule rule, Map<String, Object> params) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("ruleId",   rule.getId());
        result.put("ruleType", rule.getRuleType());

        try {
            // Resolve property mappings
            Map<String, Object> resolvedProps = resolvePropertyMappings(rule.getPropertyMappingsJson(), params);

            switch (rule.getRuleType()) {
                case "CREATE_OBJECT" -> {
                    OntologyObjectType ot = objectTypeRepo.findByName(rule.getTargetObjectTypeName())
                        .orElseThrow(() -> new NoSuchElementException("ObjectType: " + rule.getTargetObjectTypeName()));
                    OntologyObject obj = new OntologyObject();
                    obj.setObjectTypeId(ot.getId());
                    obj.setPropertiesJson(toJson(resolvedProps));
                    OntologyObject created = objectService.createObject(obj);
                    result.put("objectId", created.getId());
                    result.put("status",   "created");
                }
                case "MODIFY_OBJECT" -> {
                    Object objIdVal = params.get("objectId");
                    if (objIdVal == null) throw new IllegalArgumentException("MODIFY_OBJECT requires 'objectId' parameter");
                    Long objId = toLong(objIdVal);
                    OntologyObject existing = objectService.getObjectById(objId);
                    existing.setPropertiesJson(toJson(resolvedProps));
                    objectService.updateObjectById(objId, existing);
                    result.put("objectId", objId);
                    result.put("status",   "modified");
                }
                case "DELETE_OBJECT" -> {
                    Object objIdVal = params.get("objectId");
                    if (objIdVal == null) throw new IllegalArgumentException("DELETE_OBJECT requires 'objectId' parameter");
                    Long objId = toLong(objIdVal);
                    objectService.deleteObjectById(objId);
                    result.put("objectId", objId);
                    result.put("status",   "deleted");
                }
                case "CREATE_LINK" -> {
                    OntologyLinkType lt = linkTypeRepo.findByName(rule.getTargetLinkTypeName())
                        .orElseThrow(() -> new NoSuchElementException("LinkType: " + rule.getTargetLinkTypeName()));
                    OntologyLink link = new OntologyLink();
                    link.setLinkTypeId(lt.getId());
                    link.setSourceObjectId(toLong(params.get("sourceObjectId")));
                    link.setTargetObjectId(toLong(params.get("targetObjectId")));
                    OntologyLink created = objectService.createLink(link);
                    result.put("linkId", created.getId());
                    result.put("status", "created");
                }
                case "DELETE_LINK" -> {
                    Long linkId = toLong(params.get("linkId"));
                    objectService.deleteLink(linkId);
                    result.put("linkId", linkId);
                    result.put("status", "deleted");
                }
                default -> throw new IllegalArgumentException("Unknown ruleType: " + rule.getRuleType());
            }
        } catch (Exception e) {
            result.put("status", "failed");
            result.put("error",  e.getMessage());
        }

        return result;
    }

    private boolean evaluateCondition(String conditionJson, Map<String, Object> params) {
        try {
            Map<String, Object> condition = mapper.readValue(conditionJson, new TypeReference<>() {});
            String field    = (String) condition.get("field");
            String operator = (String) condition.get("operator");
            Object value    = condition.get("value");
            Object actual   = params.get(field);
            if (actual == null) return false;
            return switch (operator) {
                case "eq"  -> actual.toString().equals(value.toString());
                case "neq" -> !actual.toString().equals(value.toString());
                case "exists" -> true;
                default -> true;
            };
        } catch (Exception e) {
            return true;
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> resolvePropertyMappings(String mappingsJson, Map<String, Object> params) {
        if (mappingsJson == null || mappingsJson.isBlank()) return new LinkedHashMap<>();
        try {
            Map<String, Object> mappings = mapper.readValue(mappingsJson, new TypeReference<>() {});
            Map<String, Object> resolved = new LinkedHashMap<>();
            for (Map.Entry<String, Object> entry : mappings.entrySet()) {
                String val = entry.getValue().toString();
                if (val.startsWith("$parameters.")) {
                    String paramName = val.substring("$parameters.".length());
                    resolved.put(entry.getKey(), params.getOrDefault(paramName, ""));
                } else {
                    resolved.put(entry.getKey(), entry.getValue());
                }
            }
            return resolved;
        } catch (Exception e) {
            return new LinkedHashMap<>();
        }
    }

    private void triggerFunctions(Long actionTypeId, List<Map<String, Object>> ruleResults,
                                   Map<String, Object> params, OntologyActionExecution execution) {
        List<OntologyFunction> functions = functionRepo.findByActionTypeId(actionTypeId);
        for (OntologyFunction fn : functions) {
            if (!"ACTIVE".equals(fn.getStatus())) continue;
            // Functions are triggered asynchronously without blocking main transaction
            // Failure is isolated
        }
    }

    private Long toLong(Object val) {
        if (val == null) throw new IllegalArgumentException("Value is null");
        if (val instanceof Long l) return l;
        if (val instanceof Integer i) return i.longValue();
        return Long.parseLong(val.toString());
    }

    private String toJson(Object obj) {
        try { return mapper.writeValueAsString(obj); }
        catch (Exception e) { return "{}"; }
    }
}
